servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java

```diff
@@ -391,6 +506,11 @@ boolean addConnection(C connection) {
                     }
                     if (i == existing.length) {
                         break;
+                    } else if (existing.length == 1 && isExpired) {
+                        // We're closing the last connection, close the Host.
+                        // Closing the host will trigger the Host's onClose method,
+                        // which will remove the host from active hosts list.
+                        Host.this.closeAsyncGracefully().subscribe();
```

Comment:
I see that `doClose` will set `connections` to `CLOSED_ARRAY`. However, it looks like here can be a race between reading `isExpired` and updating `isExpired` + creating a new connection. Should we try to CAS `existing` with `CLOSED_ARRAY` before closing the host, WDYT?

If we set `CLOSED_ARRAY`, it guarantees no more connections can be added for this host. Therefore, we don't need to close the host gracefully because it's known that there are no connections left.

Reply:
As a matter of fact, the `doClose` method does the CAS and is called from `closeAsyncGracefully`. The purpose of closing the `Host` here is triggering the callback which will remove it from the `activeHosts` list. So we achieve two purposes at once.

The race condition that's worth considering is when a connection is currently created (c1) (let's say just a moment before `unavailable` event arrives) and the Host has just one connection (c0) open:
1. t0: Host is available, connection c1 is being created.
2. t1: SD sends unavailable event, host is marked as expired.
3. t2: Remote server closes existing connections, last one (c0) triggers closing the Host.
4. t3: New connection, a successfully established one (c1) is added to the `connections` array.

If t2 happens before t3: `addConnection` uses CAS for adding the connection and will spot if the array is already closed, rejecting the connection.
If t3 happens before t2: c0's close callback is triggered and can either see `existing.length == 1` or `existing.length == 2`. If it's `2`, nothing bad will happen. If it's `1`, `doClose` will kill the newly created connection (c1) and the Host will be removed from `activeHosts`.

In this scenario doing the CAS to `CLOSED_ARRAY` will allow us to see that in fact c0 is not the last connection, do another loop and simply remove c0 from the array.

Got it, the updated code is the following:

```
                    } else if (existing.length == 1 && isExpired) {
                        // We're closing the last connection, close the Host.
                        // Closing the host will trigger the Host's onClose method,
                        // which will remove the host from active hosts list.
                        // If a race condition appears and a new connection was added in the meantime,
                        // the CAS operation will allow for determining that and not closing the Host yet and removing
                        // the previously last connection from the array.
                        if (connectionsUpdater.compareAndSet(this, existing, CLOSED_ARRAY)) {
                            Host.this.closeAsyncGracefully().subscribe();
                        }
                    }
```